convert_results.py

```diff
@@ -0,0 +1,182 @@
+  entries = []
+  description = test.get('description', 'Unnamed test.')
+  entries.append('* {}'.format(description))
+  error_message = test.get('error_message', 'No error message.')
+  entries.append('  * {}'.format(error_message))
+  observations = test.get('observations', [])
+  entries += parse_bulleted_list(observations)
+  return '\n'.join(entries)
+
+
+def capabilities_to_text(capabilities):
+  """Returns a list of lines in a Markdown file for the given capabilities."""
+  entries = []
+  entries.append('* HID')
+  cbor = capabilities.get('cbor', False)
+  entries.append('  * CBOR: {}'.format(cbor))
+  msg = capabilities.get('msg', False)
+  entries.append('  * MSG : {}'.format(msg))
+  wink = capabilities.get('wink', False)
+  entries.append('  * WINK: {}'.format(wink))
+
+  versions = capabilities.get('versions', [])
+  if versions:
+    entries.append('* Versions')
+  entries += parse_bulleted_list(versions)
+
+  options = capabilities.get('options', [])
+  if options:
+    entries.append('* Options')
+  entries += parse_bulleted_list(options)
+
+  extensions = capabilities.get('extensions', [])
+  if extensions:
+    entries.append('* Extensions')
+  entries += parse_bulleted_list(extensions)
+
+  signature_counter = capabilities.get('signature_counter',
+                                       'No signature counter information.')
+  entries.append('* {}'.format(signature_counter))
+  return '\n'.join(entries)
+
+
+def convert(result):
+  """Converts the result file JSON content to Markdown file content."""
+  try:
+      result = json.loads(result)
+  except ValueError as _:
+    print('File was not valid JSON.')
+    return ''
+
+  for key in ['capabilities', 'device_under_test', 'tests']:
+    if key not in result:
+      print('JSON was missing the mandatory key {}.'.format(key))
+      return ''
+
+  tests = result['tests']
+  for test in tests:
+    if 'result' not in test:
+      test['result'] = 'fail'
+
+  device_under_test = result['device_under_test']
+  product_name = device_under_test.get('product_name', 'Unknown')
+  serial_number = device_under_test.get('serial_number', 'None')
+  manufacturer = device_under_test.get('manufacturer', 'None')
+  vendor_id = device_under_test.get('vendor_id', 'None')
+  product_id = device_under_test.get('product_id', 'None')
+  aaguid = device_under_test.get('aaguid', 'None')
+  device_info = [serial_number, manufacturer, vendor_id, product_id, aaguid]
+
+  passed = sum(1 if test['result'] == 'pass' else 0 for test in tests)
+  passed_test_count = result.get('passed_test_count', passed)
+  total_test_count = result.get('total_test_count', len(tests))
+  summary = summary_table(tests, passed_test_count, total_test_count)
+
+  failed_tests = []
+  for test in tests:
+    if test['result'] == 'fail':
+      failed_tests.append(test_to_string(test))
+  failed_test_text = '\n'.join(failed_tests)
+
+  capabilities = capabilities_to_text(result['capabilities'])
+
+  text = '''## {}
```

Comment:
Don't put a template inside a python file - instead you should use a templating system to separate the data from the formating. https://jinja.palletsprojects.com/en/2.11.x/

Reply:
jinja was nice, added!